koswat/strategies/order_strategy/order_strategy.py

```diff
@@ -91,28 +91,25 @@ def split_reinforcements() -> tuple[
             reverse=True,
         )
 
+        # Remove the reinforcements that are more expensive and less or equally restrictive than 1 of the others
         def check_reinforcement(
             pair: tuple[StrategyReinforcementInput, StrategyReinforcementInput],
         ) -> StrategyReinforcementInput | None:
-            # Only keep the less restrictive reinforcement if it is cheaper
             if (
-                pair[0].ground_level_surface > pair[1].ground_level_surface
-                and pair[0].base_costs < pair[1].base_costs
+                pair[0].base_costs > pair[1].base_costs
+                and pair[0].ground_level_surface >= pair[1].ground_level_surface
             ):
                 return pair[0]
             return None
 
-        # Check if the current (more expensive) reinforcement is more restrictive than the previous
-        # (the last needs to be appended as it is always kept)
-        _sorted_pairs = pairwise(_sorted + _last)
-        _kept = (
-            list(
-                filter(lambda x: x is not None, map(check_reinforcement, _sorted_pairs))
-            )
-            + _last
+        _pairs = product(_sorted, _sorted + _last)
+        _to_remove = set(
+            filter(lambda x: x is not None, map(check_reinforcement, _pairs))
         )
+        for _reinforcement in _to_remove:
+            _sorted.remove(_reinforcement)
 
-        return [x.reinforcement_type for x in _kept]
+        return [x.reinforcement_type for x in _sorted + _last]
```

Comment:
not sure why it's added now and not earlier (you are already doing a `product(_sorted, _sorted +_last)`.

Reply:
The entries in `product` are only used to determine which need to be removed. `_last` is never removed and needs to be added here.